task-context.md

```diff
@@ -0,0 +1,193 @@
+to take explicit `task::Context` arguments. `task::Context` is a type which
+provides access to
+[task-locals](https://docs.rs/futures/*/futures/macro.task_local.html) and a `Waker`.
+
+In terms of the old `futures` task model, the `Waker` is analogous to the `Task`
+returned by [`task::current`][current], and the `wake` method is analogous to
+the old [`Task::notify`][notify].
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `task::Context` interface looks like this:
+
+```rust
+mod task {
+
+  /// The context for the currently running task.
+  pub struct Context {
+
+      /// Returns a `Waker` which can be used to wake up the current task.
+      pub fn waker(&self) -> Waker { ... }
+
+      /// Runs the provided closure with the task-local pointed to by the
+      /// provided `key`.
+      ///
+      /// `LocalKey`s are acquired using the `task_local` macro.
+      pub fn with_local<T, R, F>(&mut self, key: LocalKey<T>, f: F) -> R
+          where F: FnOnce(&mut T) -> R
+      {
+          ...
+      }
+  }
+
+  /// A handle used to awaken a task.
+  pub struct Waker {
+
+      /// Wakes up the associated task.
+      ///
+      /// This signals that the futures associated with the task are ready to
+      /// be `poll`ed again.
+      pub fn wake(&self) { ... }
+  }
+
+  ...
+}
+```
+
+The modified traits will look as follows:
+
+```rust
+pub trait Future {
+    type Item;
+    type Error;
+    fn poll(&mut self, ctx: &mut Context) -> Poll<Self::Item, Self::Error>;
+}
+
+pub trait Stream {
+    type Item;
+    type Error;
+    fn poll_next(&mut self, ctx: &mut Context) -> Poll<Option<Self::Item>, Self::Error>;
+}
+
+pub trait AsyncRead {
+    type Error;
+    unsafe fn initializer(&self) -> Initializer { ... }
+
+    fn poll_read(&mut self, buf: &mut [u8], ctx: &mut Context)
+        -> Poll<usize, Self::Error>;
+
+    fn poll_vectored_read(&mut self, vec: &mut [&mut IoVec], ctx: &mut Context)
+        -> Poll<usize, Self::Error> { ... }
+}
+
+pub trait AsyncWrite {
+    type Error;
+    fn poll_write(&mut self, buf: &[u8], ctx: &mut Context)
+        -> Poll<usize, Self::Error>;
+
+    fn poll_vectored_write(&mut self, vec: &[&IoVec], ctx: &mut Context)
+        -> Poll<usize, Self::Error> { ... }
+
+    fn poll_close(&mut self, ctx: &mut Context) -> Poll<(), Self::Error>;
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The primary drawback here is an ergonomic hit, due to the need to explicitly thread through 
+a `Context` argument.
+
+For combinator implementations, this hit is quite minor.
+
+For larger custom future implementations, it remains possible to use TLS internally to recover the existing ergonomics, if desired.
+
+Finally, and perhaps most importantly, the explicit argument is never seen by users who only use
+futures via combinators or async-await syntax, so the ergonomic impact to such users in minimal.
```

Comment:
If the explicit argument is never seen by users who ony use futures via combinators or async-await syntax, then why not make the ergonomics solid for those of us who do?

Really, it seems that this proposal hinges on:

* How often do users transition from combinators only -> writing `Async` functions?
* Is it worth making that initial transition smoother at the cost of making the experience of writing `Async` functions worse?
* How much smoother will that transition actually be with an explicit task argument?

Reply:
In my experience, there's a point (and it comes pretty quickly) where you want to implement a future on you own. I'd not push that towards being an edge-case.

Then again, these are mostly futures that proxy the calls through to child futures. The biggest danger there, though, is _forgetting to make sure there's always a future polled_, which the task argument doesn't fix.

Reply:
> The biggest danger there, though, is forgetting to make sure there's always a future polled, which leaves the future group in waiting for poll. The task argument doesn't fix.

That is a good point. The explicit task argument doesn't actually statically prevent you from making the mistake that this RFC is attempting to protect you from.

Reply:
I'd question the notion that the TLS solution is more ergonomic. While it does reduce the amount of characters need to write a poll implementation that is at best a weak argument in favor of it. On the other hand I'd argue that keeping things explicit improves "ergonomics" just by virtue of being explicit.

Reply:
> I'd question the notion that the TLS solution is more ergonomic. 

Having written a *lot* of custom `impl Future`, it definitely is more ergonomic.

Reply:
@seanmonstar 
> Having written a lot of custom impl Future, it definitely is more ergonomic.

I assume there is something more than not needing to pass a context that makes this more ergonomic? If so, what is it?

Reply:
`impl Future` is explicit, than

```
struct Foo;
impl Future for Foo {}
```